### PR TITLE
Added exercise 2 functionality

### LINKED ISSUE #3: Step 2
Add next functions
- operator+
- operator==
- operator=( const char* s )
- operator[]
- add insert functions to insert  mystring or char* in certain position
- add insert functions to insert some number of symbols of mystring or char* in certain position(add additional parameter)

### PR DESCRIPTION
**Pull request by Psyhich**
Force changed to this branch to split different pull-requests for different issues
Closes #3

### CODE REVIEW

File: Exercise/src/my_string.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+void Exercise_1::CMyString::CopyString(const char *cszStringToCopy)
+{
+	std::size_t nIndex = 0;
+	for(; cszStringToCopy[nIndex] != '\0'; nIndex++)
+	{
+		m_szData[nIndex] = cszStringToCopy[nIndex];
+	}
+	m_szData[nIndex] = '\0';
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: you should add more check in this function

File: Exercise/src/my_string.cpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
+#include "my_string.h"
+#include <cstddef>
+
+std::size_t Exercise_1::CMyString::GetStringLength(const char *cszStringToCount) noexcept {
+	std::size_t nStringLength = 0;
+	char cCurrentChar = cszStringToCount[0];
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: wrong prefix.
Check style document

File: Exercise/src/my_string.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	for(; cszStringToCopy[nIndex] != '\0'; nIndex++)
+	{
+		m_szData[nIndex] = cszStringToCopy[nIndex];
+	}
+	m_szData[nIndex] = '\0';
+}
+
+void Exercise_1::CMyString::AppendToString(const char *cszStringToAppend)
+{
+
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: what if cszStringToAppend == nullptr

File: Exercise/src/my_string.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	m_szData[nIndex] = '\0';
+}
+
+void Exercise_1::CMyString::AppendToString(const char *cszStringToAppend)
+{
+	
+	size_t nTerminantIndex = GetStringLength(m_szData) - 1;
+	for(size_t nIndex = nTerminantIndex; true; nIndex++)
+	{
+		const char ccCharToAppend = cszStringToAppend[nIndex - nTerminantIndex];
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: wrong prefix

File: Exercise/src/my_string.cpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		m_szData[nIndex] = ccCharToAppend;
+		if(ccCharToAppend == '\0')
+		{
+			break;
+		}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: So, You will get m_szData without \0?

File: Exercise/src/my_string.cpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: What about size reallocating when you append the new string?

File: Exercise/src/my_string.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+void Exercise_1::CMyString::Insert(const char* cszStringToInsert, size_t nStartPosition, size_t nTimesToInsert)
+{
+	const size_t nLengthOfInserted = GetStringLength(cszStringToInsert);
+	if(nStartPosition > size() || nLengthOfInserted == 0)
+	{
+		return; // Or it is better to throw? //TODO clarify which variant is better
+	}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: print an error if parameters are not correct

File: Exercise/src/my_string.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+		m_szData[nIndex] = ccCharToAppend;
+		if(ccCharToAppend == '\0')
+		{
+			break;
+		}
+	}
+
+}
+
+void Exercise_1::CMyString::Insert(const char* cszStringToInsert, size_t nStartPosition, size_t nTimesToInsert)
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: I wanted you to add a third parameter related to number of symbols to insert/
For example if you have m_data = "trainee"
and cszStringToInsert = "_developer" and pos is last symbol
the third parameter means how many symbols from "_developer" will be inserted.
For example if 3 the result will be "trainee_de"

File: Exercise/src/my_string.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	delete[] m_szData;
+	m_szData = szNewString;
+	m_nSize = nNewSize;
+}
+
+void Exercise_1::CMyString::ReinitializeAndCopy(const char* cszStringToCopy, const std::size_t& nStringLength)
+{
+	bool shouldCleanup = m_szData != nullptr && m_nSize != nStringLength;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: prefix

Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Fixed

File: Exercise/src/my_string.cpp (hunk 2 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Fixed

File: Exercise/src/my_string.cpp (hunk 4 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Fixed all char prefixes

File: Exercise/src/my_string.cpp (hunk 3 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Added nullptr checks in all functions

File: Exercise/src/my_string.cpp (hunk 7 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Added error printing

File: Exercise/src/my_string.cpp (hunk 8 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Remade function

**Comment by Psyhich** [comment]
Test of Bot

File: Exercise/src/my_string.cpp (hunk 6 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Added reallocation to Append function

File: Exercise/src/my_string.cpp (hunk 1 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Added nullptr checks

**Comment by Psyhich** [comment]
Now, all pipeline works fine
It's time to update code and add more tests)

File: Exercise/src/my_string.cpp (hunk 5 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Fixed

**Comment by Psyhich** [comment]
Jenkins fails to publish approves on branches, so I turned off this functionality. Now it checks only pull requests

File: Exercise/src/my_string.cpp [hunk 10]
```diff
@@ -5,6 +5,10 @@
 #include "my_string.h"
 
 char* Exercise_1::CMyString::TryToAllocate(const size_t nLength) noexcept {
+	if(nLength == 0)
+	{
+		return nullptr;
+	}
 	char *pcAllocated;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:12]: Why did you add noexcept?
If you try to handle exception in the function.
If you don't want to deal with exception just add noexcept after "new" and check ptr for nullptr after allocating

Comment (VictorApopiy) [reply Exercise/src/my_string.cpp:12]: Also imitialize this ptr with nullptr

File: Exercise/src/my_string.cpp [hunk 11]
```diff
@@ -18,49 +22,143 @@ char* Exercise_1::CMyString::TryToAllocate(const size_t nLength) noexcept {
 }
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:22]: It is better to modify size member in this function.

File: Exercise/src/my_string.cpp [hunk 12]
```diff
@@ -18,49 +22,143 @@ char* Exercise_1::CMyString::TryToAllocate(const size_t nLength) noexcept {
...
+	if(cszStringToAppend == nullptr)
+	{
+		return;
+	}
+	
+	ReinitializeAndCopy(m_szData, size() + GetStringLength(cszStringToAppend) - 1);
+	
 
-	CopyString(cszCharsSequence, m_szData, m_nSize);
+	size_t nTerminantIndex = GetStringLength(m_szData) - 1;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:67]: Is it old terminationIndex before reallocating?

File: Exercise/src/my_string.cpp [hunk 13]
```diff
@@ -18,49 +22,143 @@ char* Exercise_1::CMyString::TryToAllocate(const size_t nLength) noexcept {
...
+	// Beggining multiple insertion of given strings 
+	size_t nEndPosition = nStartPosition + nCountOfCharsToInsert;
+	for(; nCurrentPosition < nEndPosition; nCurrentPosition++){
+		szNewString[nCurrentPosition] = cszStringToInsert[nCurrentPosition - nStartPosition];
+	}
+	// Appending last part
+	for(; nCurrentPosition < nNewSize; nCurrentPosition++)
+	{
+		szNewString[nCurrentPosition] = m_szData[nStartPosition + nCurrentPosition - nEndPosition];
+	}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:111]: It is better to create 1 loop with more condition blocks.

**Review by VictorApopiy — APPROVED** [review]

File: Exercise/src/my_string.cpp (hunk 10 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:12]: Fixed, by adding std::nothrow

File: Exercise/src/my_string.cpp (hunk 12 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:67]: Fixed this stupid error, now it uses m_nSize

File: Exercise/src/my_string.cpp (hunk 13 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:111]: Fixed
